code/week6/findBlue.cpp

```diff
@@ -0,0 +1,81 @@
+#include <unistd.h>
+#include <opencv2/opencv.hpp>
+#include <opencv2/core/core.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/highgui/highgui.hpp>
+
+using namespace cv;
+
+
+Mat findBlue(const Mat& frameBGR) {
+      //cv::Scalar is a container that is used to hold pixel values
+      //the values shown below are tuned by hand on example data that is
+      //gathered
+      const Scalar blue_low{78, 50, 70};     //these are the lowest HSV values
+                                             //that we accept as blue
+      const Scalar blue_high{138, 255, 255}; //these are the highest HSV
+                                             //values we accept as blue
+
+      // kernel that is an ellipse of size 11X11 pixels
+      Mat erosion_kernel_blue = getStructuringElement(MORPH_ELLIPSE, Size(11, 11));
+
+      //applying a GaussianBlur makes the image "fuzzier". This smoothes out
+      //noisy pixels
+      Mat frameBlurred;
+      GaussianBlur(frameBGR, frameBlurred, Size{7,7}, 0);
+
+      //We shift the mat which is encodes as an RGB image into an HSV image
+      Mat frameHSV;
+      cvtColor(frameBlurred, frameHSV, CV_BGR2HSV);
+
+      //We initialize an empty Mat to store all the pixels that we identify
+      //as blue. CV_8U is the image depth: each pixel is only one bit
+      Mat output_blue = Mat::zeros(200, 640, CV_8U);
+
+      //for every pixel in our frame that we define as blue, write a white
+      //pixel to the output_blue Mat
+      inRange(frameHSV, blue_low, blue_high, output_blue);
+
+      //apply an erode operation onto the filtered image. This helps us
+      //remove specks of blue that show up due to noise or glare
+      erode(output_blue, output_blue, erosion_kernel_blue);
+
+      //the logic to publish the result of the filtering in ROS would go here
+    return output_blue;
+}
+
+int main() {
+    // sets up the input stream to the default video device
+    // /dev/video0 must exist
+    VideoCapture stream1(0);
+
+    // this checks to see if a key has been pressed and then shows the image for
+    // one millisecond
+    while (waitKey(1) == 255) {
```

Comment:
>Specifically, don't worry too much about them, since none of the students will be using them. 

There was one student last year in robocup training (who left unfortunately) who used i3 (and many returning robocup team members, including will, at least partially use i3 or other tiling window managers).

There was an incoming robocup member who used emacs + unity to make his own mouseless workflow (which was what inspired me to make my own).

> boo to non-standard window managers

Assuming you're talking about unity (since it seems to be dominant at rj) unity is [officially deprecated by ubuntu](https://insights.ubuntu.com/2017/04/05/growing-ubuntu-for-cloud-and-iot-rather-than-phone-and-convergence/) and will be dropped in future ubuntu versions.

While we shouldn't prioritize alternative workflows, we should solve breaking issues as they are discovered, to avoid alienating future members. I would be happy to do this if no one else is willing to, of course. Diversity in all it's forms is important for organizations to have a balanced viewpoint.